File: python/examples/quickstart.py
"""Interpolate a quote, then find out whether you should have.

Run:  python examples/quickstart.py
"""

from __future__ import annotations

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "src"))

from fintech_linear_quote import (  # noqa: E402
    PendingInterpolator,
    gap_profile,
    interpolation_error,
    linear_quote_interpolation,
    lookahead_cost,
    status_summary,
)

P = {"instrument": "A", "venue": "X", "session_id": "S"}

QUOTES = [
    {**P, "event_time": "2026-01-01T00:00:00.000Z",
     "available_time": "2026-01-01T00:00:00.100Z", "bid": 99.0, "ask": 101.0},
    {**P, "event_time": "2026-01-01T00:00:02.000Z",
     "available_time": "2026-01-01T00:00:02.120Z", "bid": 101.0, "ask": 103.0},
    {**P, "event_time": "2026-01-01T00:00:03.000Z",
     "available_time": "2026-01-01T00:00:03.050Z", "bid": 100.0, "ask": 102.0},
]

MAX_GAP_MS = 2000


def rule(title: str) -> None:
    print(f"\n{title}\n{'-' * len(title)}")


# --- 1. the same target, asked too early and then late enough -------------- #
rule("1. The knowledge cutoff decides whether there is an answer at all")

for evaluation in ("2026-01-01T00:00:01.000Z", "2026-01-01T00:00:02.120Z"):
    row = linear_quote_interpolation(
        QUOTES,
        [{**P, "target_time": "2026-01-01T00:00:01.000Z", "evaluation_time": evaluation}],
        MAX_GAP_MS,
    )[0]
    detail = (f"bid={row['bid']}" if row["bid"] is not None
              else f"wait {row['wait_ms']}ms longer")
    print(f"  evaluated at {evaluation[-13:-1]}  ->  {row['status']:<22} {detail}")
print("  The right-hand endpoint had not arrived. There was no honest answer yet.")

# --- 2. interpolated is never observed ------------------------------------- #
rule("2. Interpolated is never observed")

targets = [
    {**P, "target_time": "2026-01-01T00:00:00.000Z", "evaluation_time": "2026-01-01T00:00:09Z"},
    {**P, "target_time": "2026-01-01T00:00:01.000Z", "evaluation_time": "2026-01-01T00:00:09Z"},
]
for row in linear_quote_interpolation(QUOTES, targets, MAX_GAP_MS):
    print(f"  {row['target_time'][-13:-1]}  {row['status']:<8} "
          f"bid={row['bid']:<7} observable={row['observable']!s:<6} "
          f"interpolated={row['interpolated']}")
print("  Only the exact, already-arrived record is an observation. The other is arithmetic.")

# --- 3. how wrong is the straight line, on this data? ---------------------- #
rule("3. Leave-one-out: how wrong is the line here?")

# The only interior quote is bracketed by 00:00 and 00:03 — a 3000ms span. At the
# 2000ms budget that bracket is skipped, which is the sweep telling you the budget
# is doing real work rather than rubber-stamping.
tight = interpolation_error(QUOTES, MAX_GAP_MS)["overall"]
print(f"  at max_gap_ms={MAX_GAP_MS}: checked {tight['checked']}, "
      f"skipped {tight['skipped']} (bracket too wide)")

stats = interpolation_error(QUOTES, 5000)["overall"]
print(f"  at max_gap_ms=5000: checked {stats['checked']} interior quote(s)")
print(f"    max absolute bid error: {stats['max_abs_bid_error']:.4f}")
print(f"    worst error in spreads: {stats['max_error_in_spreads']:.4f}")
print("  Hidden each interior quote, rebuilt it from its neighbours, compared.")

# --- 4. what did honesty cost? --------------------------------------------- #
rule("4. Lookahead a revised-file backtest would have absorbed")

early = [{**P, "target_time": "2026-01-01T00:00:01.000Z",
          "evaluation_time": "2026-01-01T00:00:01.000Z"}]
cost = lookahead_cost(QUOTES, early, MAX_GAP_MS)
print(f"  {cost['summary']['differing']}/{cost['summary']['total_targets']} targets differ")
for row in cost["rows"]:
    print(f"    honest={row['point_in_time_status']} (bid={row['point_in_time_bid']})  "
          f"revised={row['revised_status']} (bid={row['revised_bid']})  "
          f"wait={row['wait_ms']}ms")

# --- 5. where should the gap budget sit? ----------------------------------- #
rule("5. Gap profile")

profile = gap_profile(QUOTES)["overall"]
print(f"  {profile['quotes']} quotes, {profile['gaps']} gaps: "
      f"min={profile['min_ms']:.0f}ms median={profile['median_ms']:.0f}ms "
      f"max={profile['max_ms']:.0f}ms")

# --- 6. the engine that waits ---------------------------------------------- #
rule("6. Interpolation has to wait, so the engine waits")

live = PendingInterpolator(MAX_GAP_MS)
live.observe(QUOTES[0])
live.submit({**P, "target_time": "2026-01-01T00:00:01.000Z"})
print(f"  after 1 quote:  resolved={len(live.resolve())}  pending={live.pending_count} "
      f"({live.pending[0]['status']})")

live.observe(QUOTES[1])
settled = live.resolve()
print(f"  after 2 quotes: resolved={len(settled)}  pending={live.pending_count}")
print(f"  -> {settled[0]['status']} bid={settled[0]['bid']} "
      f"first_knowable_at={settled[0]['first_knowable_at']}")

# --- 7. what is this batch actually made of? ------------------------------- #
rule("7. Measurement or model?")

summary = status_summary(linear_quote_interpolation(QUOTES, targets, MAX_GAP_MS))
print(f"  {summary['observable']} observed, {summary['interpolated']} interpolated, "
      f"{summary['unresolved']} unresolved  "
      f"(observable share {summary['observable_share']:.0%})")

File: typescript/examples/quickstart.ts
/**
 * Interpolate a quote, then find out whether you should have.
 *
 * Run:  npm run example
 */

import {
  PendingInterpolator,
  gapProfile,
  interpolationError,
  linearQuoteInterpolation,
  lookaheadCost,
  statusSummary,
  type Quote,
} from "../src/index.ts";

const P = { instrument: "A", venue: "X", session_id: "S" };

const QUOTES: Quote[] = [
  { ...P, event_time: "2026-01-01T00:00:00.000Z",
    available_time: "2026-01-01T00:00:00.100Z", bid: 99.0, ask: 101.0 },
  { ...P, event_time: "2026-01-01T00:00:02.000Z",
    available_time: "2026-01-01T00:00:02.120Z", bid: 101.0, ask: 103.0 },
  { ...P, event_time: "2026-01-01T00:00:03.000Z",
    available_time: "2026-01-01T00:00:03.050Z", bid: 100.0, ask: 102.0 },
];

const MAX_GAP_MS = 2000;

const rule = (title: string) => console.log(`\n${title}\n${"-".repeat(title.length)}`);
const clock = (timestamp: string) => timestamp.slice(-13, -1);
const fixed = (value: number | null, digits = 4) =>
  value === null ? "n/a" : value.toFixed(digits);

// --- 1. the same target, asked too early and then late enough -------------- //
rule("1. The knowledge cutoff decides whether there is an answer at all");

for (const evaluation of ["2026-01-01T00:00:01.000Z", "2026-01-01T00:00:02.120Z"]) {
  const row = linearQuoteInterpolation(
    QUOTES,
    [{ ...P, target_time: "2026-01-01T00:00:01.000Z", evaluation_time: evaluation }],
    MAX_GAP_MS,
  )[0]!;
  const detail = row.bid === null ? `wait ${row.wait_ms}ms longer` : `bid=${row.bid}`;
  console.log(`  evaluated at ${clock(evaluation)}  ->  ${row.status.padEnd(22)}${detail}`);
}
console.log("  The right-hand endpoint had not arrived. There was no honest answer yet.");

// --- 2. interpolated is never observed ------------------------------------- //
rule("2. Interpolated is never observed");

const targets = [
  { ...P, target_time: "2026-01-01T00:00:00.000Z", evaluation_time: "2026-01-01T00:00:09Z" },
  { ...P, target_time: "2026-01-01T00:00:01.000Z", evaluation_time: "2026-01-01T00:00:09Z" },
];
for (const row of linearQuoteInterpolation(QUOTES, targets, MAX_GAP_MS)) {
  console.log(
    `  ${clock(row.target_time)}  ${row.status.padEnd(8)} bid=${String(row.bid).padEnd(7)} ` +
      `observable=${String(row.observable).padEnd(6)} interpolated=${row.interpolated}`,
  );
}
console.log("  Only the exact, already-arrived record is an observation. The other is arithmetic.");

// --- 3. how wrong is the straight line, on this data? ---------------------- //
rule("3. Leave-one-out: how wrong is the line here?");

// The only interior quote is bracketed by 00:00 and 00:03 — a 3000ms span. At the
// 2000ms budget that bracket is skipped, which is the sweep telling you the budget is
// doing real work rather than rubber-stamping.
const tight = interpolationError(QUOTES, MAX_GAP_MS).overall;
console.log(
  `  at maxGapMs=${MAX_GAP_MS}: checked ${tight.checked}, skipped ${tight.skipped} (bracket too wide)`,
);
const stats = interpolationError(QUOTES, 5000).overall;
console.log(`  at maxGapMs=5000: checked ${stats.checked} interior quote(s)`);
console.log(`    max absolute bid error: ${fixed(stats.max_abs_bid_error)}`);
console.log(`    worst error in spreads: ${fixed(stats.max_error_in_spreads)}`);
console.log("  Hidden each interior quote, rebuilt it from its neighbours, compared.");

// --- 4. what did honesty cost? --------------------------------------------- //
rule("4. Lookahead a revised-file backtest would have absorbed");

const cost = lookaheadCost(
  QUOTES,
  [{ ...P, target_time: "2026-01-01T00:00:01.000Z", evaluation_time: "2026-01-01T00:00:01.000Z" }],
  MAX_GAP_MS,
);
console.log(`  ${cost.summary.differing}/${cost.summary.total_targets} targets differ`);
for (const row of cost.rows) {
  console.log(
    `    honest=${row.point_in_time_status} (bid=${row.point_in_time_bid})  ` +
      `revised=${row.revised_status} (bid=${row.revised_bid})  wait=${row.wait_ms}ms`,
  );
}

// --- 5. where should the gap budget sit? ----------------------------------- //
rule("5. Gap profile");

const profile = gapProfile(QUOTES).overall;
console.log(
  `  ${profile.quotes} quotes, ${profile.gaps} gaps: min=${profile.min_ms}ms ` +
    `median=${profile.median_ms}ms max=${profile.max_ms}ms`,
);

// --- 6. the engine that waits ---------------------------------------------- //
rule("6. Interpolation has to wait, so the engine waits");

const live = new PendingInterpolator(MAX_GAP_MS);
live.observe(QUOTES[0]);
live.submit({ ...P, target_time: "2026-01-01T00:00:01.000Z" });
console.log(
  `  after 1 quote:  resolved=${live.resolve().length}  pending=${live.pendingCount} ` +
    `(${live.pending[0]!.status})`,
);

live.observe(QUOTES[1]);
const settled = live.resolve();
console.log(`  after 2 quotes: resolved=${settled.length}  pending=${live.pendingCount}`);
console.log(
  `  -> ${settled[0]!.status} bid=${settled[0]!.bid} ` +
    `first_knowable_at=${settled[0]!.first_knowable_at}`,
);

// --- 7. what is this batch actually made of? ------------------------------- //
rule("7. Measurement or model?");

const summary = statusSummary(linearQuoteInterpolation(QUOTES, targets, MAX_GAP_MS));
console.log(
  `  ${summary.observable} observed, ${summary.interpolated} interpolated, ` +
    `${summary.unresolved} unresolved  ` +
    `(observable share ${Math.round(summary.observable_share * 100)}%)`,
);
